modules/sdk-core/src/bitgo/lightning/parseAndValidateLnurlInvoice.ts

```diff
@@ -0,0 +1,99 @@
+import * as bolt11 from 'bolt11';
+import * as crypto from 'crypto';
+
+export type ParsedLightningInvoice = {
+  milliSatoshis: string;
+  /** hex-encoded */
+  paymentHash: string;
+  /** hex-encoded */
+  pubkey: string;
+  /** hex-encoded */
+  descriptionHash?: string;
+};
+
+function getNetworkForInvoice(invoice: string) {
+  if (invoice.startsWith('lntbs')) {
+    // signet invoices are not supported by our bolt11 dependency
+    // because the `tbs` prefix used for the invoice does not match the `tb`
+    // prefix used by on-chain signet addresses
+    // see: https://github.com/bitcoinjs/bolt11/pull/58#issuecomment-1106495709
+    // we can still decode the invoice however using a custom network
+    return {
+      bech32: 'tbs',
+      pubKeyHash: 0x6f,
+      scriptHash: 0xc4,
+      validWitnessVersions: [0, 1],
+    };
+  }
+
+  return undefined;
+}
+
+/**
+ * @param {ParsedLightningInvoice} invoice - a parsed lightning invoice
+ * @param {number} amount - amount intended to pay for the invoice
+ * @param {string} metadata - metadata that is used to verify the fetched invoice
+ * @throws error for invoice that does not match with amount and metadata
+ */
+export function validateLnurlInvoice(invoice: ParsedLightningInvoice, milliSatAmount: string, metadata: string): void {
+  const { milliSatoshis, descriptionHash } = invoice;
+  if (milliSatoshis !== milliSatAmount) {
+    throw new Error('amount of invoice does not match with given amount');
+  }
+
+  const hash = crypto.createHash('sha256').update(metadata).digest('hex');
+  if (descriptionHash !== hash) {
+    throw new Error('invoice h tag does not match with hash of metadata');
+  }
+}
+
+/**
+ * @param {string} coin - a type of coin for lightning
+ * @param {string} invoiceStr - a lightning invoice
+ * @return {ParsedLightningInvoice}
+ * @throws error for invalid lightning invoice
+ */
+export function parseLightningInvoice(coin: string, invoiceStr: string): ParsedLightningInvoice {
```

Comment:
The lnurl server might give us an undefined `invoiceStr` or even something unexpected like a number or object. We should probably have `invoiceStr` be `unknown` here and do basic checks to make sure it's a string.